File: sdk/python/validation/judge.py
"""LLM judge calls, baseline comparison, and confidence computation."""

from __future__ import annotations

import json
import sys
import warnings
from dataclasses import dataclass

from .config import Settings


def _validate_judge_response(result: dict) -> tuple[int, str]:
    """Clamp score 1-5, warn on malformed."""
    score = result.get("score")
    reason = result.get("reason", result.get("error", ""))
    if score is None:
        warnings.warn(f"Judge returned no score: {result}")
        return 0, str(reason)
    try:
        score = int(score)
    except (TypeError, ValueError):
        warnings.warn(f"Judge returned non-integer score: {score}")
        return 0, str(reason)
    if score < 1 or score > 5:
        warnings.warn(f"Judge score {score} out of range, clamping to 1-5")
        score = max(1, min(5, score))
    return score, str(reason)


def call_judge(settings: Settings, prompt: str) -> dict:
    try:
        from openai import OpenAI
    except ImportError:
        print("ERROR: openai package not installed. uv add openai", file=sys.stderr)
        sys.exit(1)

    client = OpenAI(api_key=settings.openai_api_key)
    resp = client.chat.completions.create(
        model=settings.judge_model,
        messages=[{"role": "user", "content": prompt}],
        temperature=0,
        max_tokens=200,
        response_format={"type": "json_object"},
    )
    text = resp.choices[0].message.content.strip()

    try:
        return json.loads(text)
    except json.JSONDecodeError:
        return {"error": text}


def judge_individual(settings: Settings, prompt: str, output: str) -> tuple[int, str]:
    max_chars = settings.judge_max_output_chars
    judge_prompt = f"""You are evaluating an AI agent's output. The agent was given this task:
"{prompt}"

The agent produced this output:
"{output[:max_chars]}"

Rate the output on a scale of 1-5:
1 = Completely wrong, irrelevant, or empty
2 = Partially relevant but mostly incorrect or incomplete
3 = Relevant but missing key elements
4 = Good response, addresses the task well
5 = Excellent, fully addresses the task

Respond with ONLY a JSON object: {{"score": N, "reason": "brief explanation"}}"""

    result = call_judge(settings, judge_prompt)
    return _validate_judge_response(result)


def judge_comparison(
    settings: Settings, prompt: str, baseline_output: str, candidate_output: str
) -> tuple[int, str]:
    """Compare candidate output against baseline, scored 1-5 on task correctness."""
    max_chars = settings.judge_max_output_chars
    judge_prompt = f"""You are comparing two AI agent outputs for the same task. The task was:
"{prompt}"

BASELINE output:
"{baseline_output[:max_chars]}"

CANDIDATE output:
"{candidate_output[:max_chars]}"

Rate the CANDIDATE relative to the BASELINE on task correctness (1-5):
5 = Both correctly address the task, candidate equally valid
4 = Candidate addresses task well, minor completeness differences
3 = Candidate partially addresses task, misses key elements baseline covered
2 = Candidate attempts task but significant parts wrong
1 = Candidate fails the task or irrelevant

Different-but-valid approaches should score high. Judge task correctness, not surface similarity.

Respond with ONLY a JSON object: {{"score": N, "reason": "brief explanation"}}"""

    result = call_judge(settings, judge_prompt)
    return _validate_judge_response(result)


@dataclass
class JudgeState:
    """Mutable state tracked across judge calls."""

    call_count: int = 0
    cache_hits: int = 0

File: sdk/typescript/src/validation/judge.ts
import type { AgentResult } from '../types.js';
import type { JudgeConfig } from './config.js';

/**
 * Rubric criterion for LLM judge evaluation.
 */
export interface JudgeRubric {
  name: string;
  description: string;
  weight?: number;
}

/**
 * Result from the LLM judge.
 */
export interface JudgeResult {
  passed: boolean;
  scores: Record<string, number>;
  weightedAverage: number;
  reasoning: Record<string, string>;
}

/**
 * Build the judge prompt.
 */
function buildJudgePrompt(
  result: AgentResult,
  rubrics: JudgeRubric[],
  maxOutputChars: number,
): string {
  const outputStr = JSON.stringify(result.output).slice(0, maxOutputChars);
  const rubricLines = rubrics
    .map(
      (r, i) =>
        `${i + 1}. ${r.name} (weight: ${r.weight ?? 1}): ${r.description}`,
    )
    .join('\n');

  return `You are an AI judge evaluating an agent's output quality.

## Agent Output
${outputStr}

## Agent Status
Status: ${result.status}
Finish Reason: ${result.finishReason}

## Rubrics
Score each rubric on a 1-5 scale (1=poor, 5=excellent):
${rubricLines}

## Response Format
Respond ONLY with valid JSON:
{
  "scores": { "rubric_name": <number>, ... },
  "reasoning": { "rubric_name": "<explanation>", ... }
}`;
}

/**
 * Compute weighted average of scores.
 */
function computeWeightedAverage(
  scores: Record<string, number>,
  rubrics: JudgeRubric[],
): number {
  let totalWeight = 0;
  let weightedSum = 0;

  for (const r of rubrics) {
    const weight = r.weight ?? 1;
    const score = scores[r.name] ?? 0;
    totalWeight += weight;
    weightedSum += score * weight;
  }

  return totalWeight > 0 ? weightedSum / totalWeight : 0;
}

/**
 * Judge an agent result against rubric criteria using an LLM.
 */
export async function judgeResult(
  result: AgentResult,
  rubrics: JudgeRubric[],
  config: JudgeConfig,
  options?: { endpoint?: string; apiKey?: string; passThreshold?: number },
): Promise<JudgeResult> {
  const maxOutputChars = config.maxOutputChars ?? 3000;
  const passThreshold = options?.passThreshold ?? 3.5;
  const endpoint =
    options?.endpoint ?? 'https://api.openai.com/v1/chat/completions';
  const apiKey = options?.apiKey ?? '';
  const prompt = buildJudgePrompt(result, rubrics, maxOutputChars);

  try {
    const response = await fetch(endpoint, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        ...(apiKey ? { Authorization: `Bearer ${apiKey}` } : {}),
      },
      body: JSON.stringify({
        model: config.model,
        messages: [{ role: 'user', content: prompt }],
        max_tokens: config.maxTokens ?? 300,
        temperature: 0,
      }),
    });

    if (!response.ok) {
      throw new Error(`Judge API returned ${response.status}`);
    }

    const data = (await response.json()) as {
      choices?: Array<{ message?: { content?: string } }>;
    };
    const content = data.choices?.[0]?.message?.content ?? '';

    const jsonMatch = content.match(/\{[\s\S]*\}/);
    if (!jsonMatch) {
      return defaultJudgeResult(rubrics, passThreshold);
    }

    const parsed = JSON.parse(jsonMatch[0]) as {
      scores?: Record<string, number>;
      reasoning?: Record<string, string>;
    };

    const scores: Record<string, number> = {};
    const reasoning: Record<string, string> = {};

    for (const r of rubrics) {
      scores[r.name] = parsed.scores?.[r.name] ?? 3;
      reasoning[r.name] = parsed.reasoning?.[r.name] ?? '';
    }

    const weightedAverage = computeWeightedAverage(scores, rubrics);

    return {
      passed: weightedAverage >= passThreshold,
      scores,
      weightedAverage,
      reasoning,
    };
  } catch {
    return defaultJudgeResult(rubrics, passThreshold);
  }
}

/**
 * Default judge result when LLM is unavailable.
 */
function defaultJudgeResult(
  rubrics: JudgeRubric[],
  passThreshold: number,
): JudgeResult {
  const scores: Record<string, number> = {};
  const reasoning: Record<string, string> = {};

  for (const r of rubrics) {
    scores[r.name] = 3;
    reasoning[r.name] = 'Default score (judge unavailable)';
  }

  const weightedAverage = computeWeightedAverage(scores, rubrics);

  return {
    passed: weightedAverage >= passThreshold,
    scores,
    weightedAverage,
    reasoning,
  };
}
